Fix parsing errors

bryan-hoang opened the pull request:
I didn't thoroughly test the plugin on a large code base, which led to spurious errors from the plugin failing to parse edge cases (e.g., failing to parse `const path = require('node:path');`). To address the issues, I decided to improve the testing DX by switching from `mocha` to `vitest` before moving on to tackling the issues.

In addition to adding multiple test cases and CI checks, I opted to use https://github.com/eslint/eslint/blob/main/lib/rules/brace-style.js as a reference to modify it to fit the brace style of the company's style guide.

bryan-hoang committed:
ci: add workflow to lint and test the project

bryan-hoang committed:
docs(brace-style): update docs

Fixes lint issues & adds relevant links/examples.

bryan-hoang committed:
refactor(brace-style): fix lint issues

In additional to switching to `messageId`, additional lint rules are
also added and followed.

bryan-hoang committed:
test(brace-style): add failing test case

Encountered when trying to use the plugin on a large code base.

bryan-hoang committed:
chore{package.json}: update engines, license & files

Updates the minimum node version to 18 to address the EOL for security
updates of Node 16 on 2023-09-11.

Commands ran to update the `package*.json` files:

```console
npm pkg set engines.node='>=18'
npm install --package-lock-only
```

Also fixed audit issues with `npm audit fix`.

Refs: https://nodejs.dev/en/about/releases/

bryan-hoang committed:
build(npm): switch to forked eslint-plugin-node

The forked version is maintained.

bryan-hoang committed:
chore: switch from mocha to vitest

Mainly for a better DX.

bryan-hoang committed:
chore(package.json): add meta data fields

bryan-hoang committed:
docs(README): update & comply w/ standard readme spec

bryan-hoang committed:
fix: fix edge cases missed

Added a lot more test cases and refactored the custom rule to be based
on the core `brace-style` rule from ESLint.

bryan-hoang committed:
feat: add recommended config

Makes it easier to configure the plugin.

bryan-hoang committed:
chore: add `@distributive` eslint config

In addition to adding the config, lint issues were addressed.

bryan-hoang committed:
chore: get exported meta data from `package.json`

Make sure `package.json` is used as the single source of truth for the
package's name & version.

bryan-hoang commented:
@j0ashm Feel free to skim it to see the changes if ya have the time. I should've done a better job testing the plugin on our monorepo 😅.

j0ashm commented:
@bryan-hoang Out of curiosity, what kind of edge cases were missed? Reading the issue, I'm not quite sure where/why `node:path` comes into play as a parsing error?

j0ashm commented:
I do remember running the plug-in on the DCP monorepo and it didn't throw up any errors... maybe I had it misconfigured?

bryan-hoang commented:
> @bryan-hoang Out of curiosity, what kind of edge cases were missed? Reading the issue, I'm not quite sure where/why `node:path` comes into play as a parsing error?

@j0ashm If you're on the `main` branch and add `const path = require('node:path');` as a test cast, the test will fail. e.g., apply [test-case.patch](https://github.com/Distributive-Network/eslint-plugin-distributive/files/12450285/test-case.patch) and run the tests on `main`.

> I do remember running the plug-in on the DCP monorepo and it didn't throw up any errors... maybe I had it misconfigured?

After linking the plugin & config in, I essentially ran `npx -- eslint --ignore-path=.gitignore --ext=js,simple,tap,simple.failing,tap.failing .`, and ran into the error on `attic/bank/bin/bank.js:3`, which is the failing test case.

bryan-hoang committed:
build(npm): Bump minimum peer dep of eslint to 8.40.0

The plugin uses `context.souceCode` which was only recently added in May
this year.

If the version of the peer dependency isn't satisfied, the plugin breaks
when the property is undefined.

Refs: https://eslint.org/blog/2023/05/eslint-v8.40.0-released/#highlights

edeluzio commented on .eslintrc.js:
```diff
@@ -1,21 +1,40 @@
-      files: ["tests/**/*.js"],
-      env: { mocha: true },
+  overrides: [ {
+    files: ['**/*.mjs'],
```
Just wanna make sure this was intentional, switching overrides from .js to .mjs. Why do we do this?

bryan-hoang replied:
Yep! Since the [`"types"`](https://nodejs.org/api/packages.html#type) field of the `package.json` file is omitted, all `.js` files by default are CommonJS modules when loaded by Node. But Vitest's config file format has to be an ECMAScript Module, so the `.mjs` extension tells Node.js to load it as an ES module.

ESLint needs to be configured to understand that behaviour, unfortunately. It'd be nice if it understood that behaviour by default.

edeluzio commented on docs/rules/brace-style.md:
```diff
@@ -0,0 +1,62 @@
+
+## Further Reading
+
+- <https://github.com/Distributive-Network/eslint-config#braces>
```
(Nitpick) Page has two of these links, one at the top and one here at the bottom. I'd vote to choose one of them, unless there's a good reason for having both

bryan-hoang replied:
@edeluzio I've made the second link refer to the README in general, rather than the same section linked twice. I'd like to keep the first link, as it's meant to explain where the rule comes from, whereas further reading would be for folks interested in learning more, I suppose.

For whatever reason, GitHub isn't updating this PR with my recent pushes :/

edeluzio reviewed (commented):
A couple nitpicks, but overall LGTM. Running without problems locally

edeluzio commented:
Good change with the link, it's minor but better IMO.
LGTM!

edeluzio reviewed (commented):
LGTM!

eroosenmaallen reviewed (approved):
Looks right
